src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEReader.cs

```diff
@@ -716,5 +717,230 @@ internal static unsafe ImmutableArray<byte> DecodeEmbeddedPortablePdbDebugDirect
+                ExceptionDispatchInfo.Capture(errorToReport).Throw();
+            }
+
+            return null;
+        }
+
+        private void ReadPortableDebugTableEntries(out DebugDirectoryEntry codeViewEntry, out DebugDirectoryEntry embeddedPdbEntry)
+        {
+            codeViewEntry = default(DebugDirectoryEntry);
+            embeddedPdbEntry = default(DebugDirectoryEntry);
+
+            foreach (DebugDirectoryEntry entry in ReadDebugDirectory())
+            {
+                if (entry.IsPortableCodeView)
+                {
+                    codeViewEntry = entry;
+                }
+                else if (entry.Type == DebugDirectoryEntryType.EmbeddedPortablePdb)
+                {
+                    embeddedPdbEntry = entry;
+                }
+            }
+        }
+
+        private MetadataReaderProvider OpenCodeViewPortablePdb(DebugDirectoryEntry codeViewEntry, string peImageDirectoryOpt, Func<string, Stream> pdbFileStreamProvider, ref Exception errorToReport)
+        {
+            CodeViewDebugDirectoryData data;
+
+            try
+            {
+                data = ReadCodeViewDebugDirectoryData(codeViewEntry);
+            }
+            catch (Exception e) when (e is BadImageFormatException || e is IOException)
+            {
+                errorToReport = errorToReport ?? e;
+                return null;
+            }
+
+            if (data.Age == 1)
+            {
+                // not a portable code view:
+                return null;
+            }
+
+            var id = new BlobContentId(data.Guid, codeViewEntry.Stamp);
+            string collocatedPdbPath = null;
+
+            try
+            {
+                if (peImageDirectoryOpt != null)
+                { 
+                    collocatedPdbPath = Path.Combine(peImageDirectoryOpt, Path.GetFileName(data.Path));
+                }
+            }
+            catch
+            {
+                errorToReport = errorToReport ?? new BadImageFormatException(SR.InvalidCodeViewPdbPath);
+            }
+
+            return
+                OpenPortablePdbFile(collocatedPdbPath, id, pdbFileStreamProvider, ref errorToReport) ??
+                OpenPortablePdbFile(data.Path, id, pdbFileStreamProvider, ref errorToReport);
+        }
+
+        private MetadataReaderProvider OpenPortablePdbFile(string pathOpt, BlobContentId id, Func<string, Stream> pdbFileStreamProvider, ref Exception errorToReport)
+        {
+            if (pathOpt == null)
+            {
+                return null;
+            }
+
+            MetadataReaderProvider result = null;
+            MetadataReaderProvider candidate = null;
+            try
+            {
+                Stream pdbStream;
+
+                try
+                {
+                    pdbStream = pdbFileStreamProvider(pathOpt);
+                }
+                catch (FileNotFoundException)
```

Comment:
Also guard with File.Exists to avoid first chance exception when debugging fully expected case?